File: SourceRecon/CalculateSensitivity.cpp
// standard libraries
#include <iostream>
#include <filesystem>
#include <chrono>

// ROOT libraries
#include "TChain.h"
#include "TH2D.h"
#include "TFile.h"
#include "TCanvas.h"

// util libraries
#include "Options.h"
#include "CalcSenseUtils.h"
#include "GenBackCountsUtils.h"

int main(int argc, char** argv){
// Parser logic adapted from https://github.com/wgseligman/GramsSim/tree/master/util

    auto options = util::Options::GetInstance();
    auto result = options->ParseOptions(argc, argv, "Sensitivity");

    // Abort if we couldn't parse the job options.
    if (!(result)){
        std::cerr << "ABORT: File " << __FILE__ << " Line " << __LINE__ << " " 
                << std::endl 
                << "Aborting job due to failure to parse options"
                << std::endl;
        return -1;
    }

    // Check for help and verbose flag
  bool help;
  options->GetOption("help",help);
  if (help) {
    options->PrintHelp();
      return 0;
  }

  bool verbose;
  options->GetOption("verbose",verbose);

  // Create basename of background histograms
  std::string BackgroundPath;
  std::string BackgroundBase;
  std::string AbsBackRootName;
  bool BackPathCheck = options->GetOption("BackgroundCountFolder", BackgroundPath);
  bool BackBaseCheck  = options->GetOption("BackgroundBaseName", BackgroundBase);
  if(BackBaseCheck && BackBaseCheck){
    // Sanity check that this file format works (doesn't mean it's the correct file but that is a pebkac)
    AbsBackRootName = BackgroundPath+"/"+BackgroundBase;
    // Check if the 0.root file exists
    if(!std::filesystem::exists(AbsBackRootName+"0.root")){
      std::cerr << AbsBackRootName+"0.root" << " doesn't exist" << std::endl;
      return -1;
    }
  }
  else{
    std::cerr << "Invalid background folder or background basename" << std::endl;
    return -1;
  }

  // Create basename of source cones
  std::string SourcePath;
  std::string SourceBase;
  std::string AbsSourceRootName;
  bool SourcePathCheck = options->GetOption("SourceConesFolder", SourcePath);
  bool SourceBaseCheck  = options->GetOption("SourceConesBaseName", SourceBase);
  if(SourceBaseCheck && SourceBaseCheck){
    // Sanity check that file exists (doesn't mean it's the correct file but that is a pebkac)
    AbsSourceRootName = SourcePath+"/"+SourceBase;
    if(!std::filesystem::exists(AbsSourceRootName+"0.root")){
      std::cerr << AbsSourceRootName+"0.root" << " doesn't exist" << std::endl;
      return -1;
    }
  }
  else{
    std::cerr << "Invalid Source folder or Source basename" << std::endl;
    return -1;
  }

  // Read in NPts to generate per Compton cone
    int NPts;
    bool successN = options->GetOption("NPts",NPts);
    if (!(successN)){
      std::cerr << "Invalid Number of Points per cone" << std::endl;
      return -1;
    }
    else{
      if( (NPts <= 0)){
        std::cerr << "Invalid Number of points per cone" << std::endl;
        return -1;
      }
    }

    // Determine number of events per batch job of source.
    // This is a workaround consisting of parsing a string and fact checking that it can be a positive long
    // I really didn't want to try and template Options.h
    std::string TempNEvents;
    long SourceEventsPerJob;
    bool successNEvents = options->GetOption("SourceEventsPerJob",TempNEvents);
    if (!(successNEvents)){
      std::cerr << "Invalid Number of total events" << std::endl;
      return -1;
    }
    else{
        // try catch to see if we can convert to long
      try{
        SourceEventsPerJob = std::stol(TempNEvents);
      }
      catch(...){
        std::cerr << "Can't convert SourceEventsPerJob  to long "  << std::endl;
        return -1;
      }
    // We need SourceEventsPerJob to be strictly positive
      if(SourceEventsPerJob<=0){
        std::cerr << "Number of total events needs be positive and non-zero" << std::endl;
        return -1;
      }
    }

    // Exposure time is strictly positive
    double ExposureTime;
    bool CheckExposure = options->GetOption("ExposureTime",ExposureTime);
    if(CheckExposure){
        if(ExposureTime<=0){
            std::cerr << "Invalid exposure time. needs to be strictly positive. Units of seconds" << std::endl;
            return -1;
        }
    }
    else{
        std::cerr << "Invalid exposure time. needs to be strictly positive" << std::endl;
        return -1;
    }

    // Energy of source is strictly positive
    double SourceEnergy;
    bool CheckSourceEnergy = options->GetOption("SourceEnergy",SourceEnergy);
    if(CheckSourceEnergy){
        if(SourceEnergy<=0){
            std::cerr << "Invalid Source Energy. needs to be strictly positive. Units MeV" << std::endl;
            return -1;
        }
    }
    else{
        std::cerr << "Invalid Source Energy. needs to be strictly positive" << std::endl;
        return -1;
    }

  // Open up Effective Area histogram
  std::string EffectiveAreaPath;
  bool EffAPathCheck = options->GetOption("EffectiveAreaRoot", EffectiveAreaPath);
  if(EffAPathCheck){
    if(std::filesystem::exists(EffectiveAreaPath)){
      try{
        // Make sure that we can open up file and read contents
        std::unique_ptr<TFile> EffAFile(TFile::Open(EffectiveAreaPath.c_str(), "READ"));
        TH2D* EffA= (TH2D*)EffAFile->Get("EffArea");
        if(EffA==NULL){
          throw 1;
        }
      }
      catch(...){
        std::cerr << "Couldn't open Effective Area histogram "  << std::endl;
        return -1;
      }
    }
    else{
      std::cerr << "Invalid Effective Area Path" << std::endl;
      return -1;
    }
  }
  else{
    std::cerr << "Invalid Effective Area Path" << std::endl;
    return -1;
  }

  // Read in total number of batches
  int batches;
  bool CheckBatch = options->GetOption("Batches",batches);
  if(CheckBatch){
    if(batches < 0){
      std::cerr << "Batches need to be a strictly positive number" << std::endl;
      return -1;
    }
  }
  else{
    std::cerr << "Invalid number of batches" << std::endl;
    return -1;
  }

    std::cout << "Reading in Background..." << std::endl;

  // Load in skeleton of BackgroundCounts from the first background count histogram
  TH1D* AggCounts;
  std::string seed = AbsBackRootName+"0.root";

  std::unique_ptr<TFile> Seed(TFile::Open(seed.c_str(), "READ"));
  AggCounts = (TH1D*)Seed->Get("Counts");
  if(AggCounts==NULL){
    std::cerr << "Couldn't read Counts histogram in " << seed << std::endl;
    return -1;
  }
  // Create histogram. We now have the correct binnings.
  AggCounts->Reset();

  // Get bin number of region of background we are comparing against
  int TargetBin = ExtractBinNum(SourceEnergy, AggCounts);

  // Aggregate all the sky maps that exist
  bool CheckAgg = ReadBackgroundCounts(AbsBackRootName,batches,AggCounts,verbose);

  // Calculate mean and std. dev. assuming Poisson stats
  double mu = AggCounts->GetBinContent(TargetBin);
  double sigma = AggCounts->GetBinError(TargetBin);
  // Calculates 3 and 5 sigma thresholds
  double ThreeSigmaThreshold = mu+3*sigma;
  double FiveSigmaThreshold = mu+5*sigma;

  std::cout << "Mean: " << mu << '\t' << "Sigma: " << sigma << std::endl;

  // Place holder variables
  Long64_t TotalRecordedPhotons = 0.0;
  Long64_t ThreeSigmaRecordedPhotons = 0.0;
  Long64_t FiveSigmaRecordedPhotons = 0.0;
  double TotalCounts = 0.0;
  // Flags to make sure that I only write the number of photons at each threshold only once
  bool ThreeSigmaFlag = false;
  bool FiveSigmaFlag = false;

  std::cout << "Calculating Sensitivity" << std::endl;
  for(int i=0; i< batches; ++i){
    // Try and read in source file
    std::string id = std::to_string(i);
    std::string path = AbsSourceRootName+id+".root";
    if(!std::filesystem::exists(path)){
      if(verbose){
          std::cerr << path << " could not be found " << std::endl;
      }
      continue;
    }
    // Get the number of cones that lie within the mask
    std::unique_ptr<TFile> Current(TFile::Open(path.c_str(), "READ"));
    TH1D* CurrentCounts = (TH1D*)Current->Get("Counts");
    if(CurrentCounts==NULL){
      std::cerr << "Couldn't read Counts histogram in" << path << std::endl;
      continue;
    }
    TotalCounts += CurrentCounts->GetBinContent(CurrentCounts->FindBin(SourceEnergy));
    // Keep a running tally of how many photons you need to observe TotalCounts number of cones
    TotalRecordedPhotons += SourceEventsPerJob;
    if(verbose){
      std::cout << "FileNum: "<< i  << " Counts: " << TotalCounts << " Thresholds: " << ThreeSigmaThreshold << '\t' << FiveSigmaThreshold << std::endl;
    }
    // If you hit the 3 sigma threshold, you save how many photons where needed to generate the current TotalCounts number of cones
    if((TotalCounts >= ThreeSigmaThreshold) && !(ThreeSigmaFlag)){
      ThreeSigmaFlag = true;
      ThreeSigmaRecordedPhotons = TotalRecordedPhotons;
    }
    // Ditto with 5 sigma, but now you break out of loop since 5 sigma is good enough
    if((TotalCounts >= FiveSigmaThreshold) && !(FiveSigmaFlag)){
      FiveSigmaFlag = true;
      FiveSigmaRecordedPhotons = TotalRecordedPhotons;
      break;
    }
  }

  // We assume that the number of photons is proportional to the number of cones generated
  // The proportionality constant might be different for each energy, but we only consider 1 energy at a time, so it's OK
  // as an example: NPhotonsThreeSigma/FiveSigmaRecordedPhotons = ThreeSigmaThresholdCones/TotalCones. Rearrange for NPhotonsThreeSigma
  double NPhotonsThreeSigma = (ThreeSigmaThreshold)/(TotalCounts)*((double) FiveSigmaRecordedPhotons);
  double NPhotonsFiveSigma = (FiveSigmaThreshold)/(TotalCounts)*((double) FiveSigmaRecordedPhotons);

  // Get the effective area at this energy
  std::unique_ptr<TFile> EffAFile(TFile::Open(EffectiveAreaPath.c_str(), "READ"));
  TH1D* EffA= (TH1D*) EffAFile->Get("EffArea");

  double SourceEffA = ExtractEffArea(SourceEnergy, EffA);

  // Conversion rate of MeV to ergs because astronomers still user ergs
  const double MeVtoErgs = 1.60218E-6;

  // Convert number of photons to sensitivity (units of ergs/(cm^2*s))
  // Multiply number of photons by source energy, convert to ergs, and divide by effective area nd exposure time
  double ThreeSigmaDiffSensitivity = NPhotonsThreeSigma*SourceEnergy*MeVtoErgs/(SourceEffA*ExposureTime);
  double FiveSigmaDiffSensitivity = NPhotonsFiveSigma*SourceEnergy*MeVtoErgs/(SourceEffA*ExposureTime);
  // Another way of representing sensitivity is photons/cm^2/s.
  double ThreeSigmaPhotonSensitivity = NPhotonsThreeSigma/(SourceEffA*ExposureTime);
  double FiveSigmaPhotonSensitivity = NPhotonsFiveSigma/(SourceEffA*ExposureTime);

  // Printing to console
  std::cout << "At a source energy of "<< SourceEnergy << " MeV and exposure time of " << ExposureTime << " seconds , the sensitivity is (in ergs*cm^-2*s^-1):" << std::endl;
  std::cout << "\tThree Sigma: " << ThreeSigmaDiffSensitivity << std::endl;
  std::cout << "\tFive Sigma: " << FiveSigmaDiffSensitivity << std::endl;
  std::cout << "In photons/cm^2/s:" <<std::endl;
  std::cout << "\tThree Sigma: " << ThreeSigmaPhotonSensitivity << std::endl;
  std::cout << "\tFive Sigma: " << FiveSigmaPhotonSensitivity << std::endl;

  return 0;
}

File: CalculateSensitivity.py
import tomllib
import argparse
import sys
import os

import ROOT
ROOT.gROOT.SetBatch(True)

def ExtractBinNum(SourceEnergy, Histogram):
## Function that extracts bin number of a particular energy, with error checks at boundaries
    nbins = Histogram.GetNbinsX()
    XAxis = Histogram.GetXaxis()
## See if SourceEnergy is higher than upper bound
    upper_bound = XAxis.GetBinCenter(nbins)
    if(SourceEnergy > upper_bound):
        print("Source energy to high, defaulting to highest bin")
        return nbins-1
## Same for lower bound
    lower_bound = XAxis.GetBinLowEdge(1)
    if(SourceEnergy<= lower_bound):
        print("Source energy too low. Defaulting to lowest bin...")
    return Histogram.FindBin(SourceEnergy)

def ReadBackgroundCounts(BackgroundFileLocation ,  BackgroundSkyMapBase, nbatches):
## Aggregate all background files counts into one histogram
## Skips files if they are not present
    first_file = os.path.join(BackgroundFileLocation,BackgroundSkyMapBase+"_0.root")
    first_file_temp = ROOT.TFile.Open(first_file,"READ")
    AggCounts = first_file_temp.Get("Counts")
    AggCounts.Reset()
    for i in range(nbatches):
        file_path = os.path.join(BackgroundFileLocation,BackgroundSkyMapBase+"_"+str(i)+".root")
        if not os.path.exists(file_path):
            print(file_path+" doesn't exists. Skipping...")
            continue
        Current = ROOT.TFile.Open(file_path,"READ")
        AddonHist = Current.Get("Counts")
        if not AddonHist:
            print("Couldn't read Counts histogram in "+ file_path+" . Skipping...")
            continue
        AggCounts.Add(AddonHist)
    Output = AggCounts.Clone()
    Output.SetDirectory(0)
    return Output

if __name__=="__main__":
## MevtoERgs conversion
    MeVtoErgs = 1.60218E-6
    parser = argparse.ArgumentParser(
                        prog='GenCondorJobs',
                        description='Generates .sh and .cmd files for condor jobs to send to Nevis cluster',
                        epilog='NOTE: Effective Area root file must be in same directory as '+sys.argv[0])
    parser.add_argument("config",help="Path to .ini config file")
## Parse toml file
    args = parser.parse_args()
    try:
        with open(args.config,'rb') as f:
            config = tomllib.load(f)
    except:
        print("Couldn't read"+ args.config)
        sys.exit()

## Read in and validate input parameters
    SourceEnergy = float(config["Source"]["SourceEnergy"])
    if(SourceEnergy <=0):
        print("Source energy must be positive")
        sys.exit()

    SourceEventsPerJob = int(config["Source"]["SourceEventsPerJob"])
    if(SourceEventsPerJob <=0):
        print("Number of jobs per batch must be is positive")
        sys.exit()

    SourceBatches = int(config["Source"]["SourceBatches"])
    if(SourceBatches <=0):
        print("Number of source batches must be is positive")
        sys.exit()

    BackBatches = int(config["Background"]["BackgroundBatches"])
    if(BackBatches <=0):
        print("Number of background batches must be is positive")
        sys.exit()

    ExposureTime = float(config["Background"]["SkyMap"]["ExposureTime"])
    if(ExposureTime <=0):
        print("Background exposure time must be is positive")
        sys.exit()

    Scaling = float(config["Sensitivity"]["TimeScaling"])
    if(Scaling <=0):
        print("Time scaling must be is positive")
        sys.exit()
## EffArea file reading
    print("Reading in Effective Area...")
    EffAreaRoot = config["CalcEffArea"]["OutputFileName"]+".root"
    try:
        EffAreaF = ROOT.TFile.Open(EffAreaRoot,"READ")
        EffArea = EffAreaF.Get("EffArea")
    except:
        print("Couldn't read Effective Area file")
        sys.exit()
## Background file reading
    print("Reading in Background...")
    BackgroundFileLocation = os.path.join(config["General"]["output_directory"], "Background", "SkyMap")
    BackgroundSkyMapBase = config["Background"]["SkyMap"]["SkyMapOutput"]
    FirstBack = os.path.join(BackgroundFileLocation,BackgroundSkyMapBase+"_0.root")
#    try:
    FirstBackRootF = ROOT.TFile.Open(FirstBack,"READ")
    FirstBackRoot = FirstBackRootF.Get("Counts")
#    except:
#        print("Couldn't read "+ FirstBack)
#        sys.exit()
## Extract background counts and calculate 3 sigma thresholds
    bin_num = ExtractBinNum(SourceEnergy,FirstBackRoot)
    AggCounts = ReadBackgroundCounts(BackgroundFileLocation ,  BackgroundSkyMapBase, BackBatches)
    mu = AggCounts.GetBinContent(bin_num)
    sigma = AggCounts.GetBinError(bin_num)
    ThreeSigmaThreshold = mu+3*sigma
    FiveSigmaThreshold = mu+5*sigma
    print("Mean: "+ str(mu)+ "Sigma: "+ str(sigma)+"\n")
# Read in source files to determine sensitivity
    SourceFileLocation = os.path.join(config["General"]["output_directory"], "Source", "SkyMap")
    SourceSkyMapBase = config["Source"]["SkyMap"]["SkyMapOutput"]
    TotalRecordedPhotons = 0
    ThreeSigmaRecordedPhotons = 0
    FiveSigmaRecordedPhotons = 0
    TotalCounts = 0.0
    ThreeSigmaFlag = False
    FiveSigmaFlag = False
    print("Calculating Sensitivity...")
    for i in range(SourceBatches):
# Skip Source files if they don't exist
        source_file_path = os.path.join(SourceFileLocation, SourceSkyMapBase+"_"+str(i)+".root")
        if not os.path.exists(source_file_path):
            print(source_file_path+" doesn't exists. Skipping...")
            continue
# Add to number of observed counts in each file to counter, and keep track of when sigma tresholds are met
        Current = ROOT.TFile.Open(source_file_path,"READ")
        CurrentCounts = Current.Get("Counts")
        TotalCounts += CurrentCounts.GetBinContent(CurrentCounts.FindBin(SourceEnergy))
        TotalRecordedPhotons += SourceEventsPerJob
        if ((TotalCounts > ThreeSigmaThreshold) and  not (ThreeSigmaFlag)):
            ThreeSigmaFlag = True
            ThreeSigmaRecordedPhotons = TotalRecordedPhotons
        if ((TotalCounts > FiveSigmaThreshold) and  not (FiveSigmaFlag)):
            FiveSigmaFlag = True
            FiveSigmaRecordedPhotons = TotalRecordedPhotons
            break
# Scaling time and effective area at source energy
    EffectiveTime = ExposureTime*Scaling
    SourceEffArea = EffArea.GetBinContent(ExtractBinNum(SourceEnergy,EffArea))
    Denom = EffectiveTime*SourceEffArea
# Calculate Number of photons needed to hit a sensitivity threshold
    NPhotonsThreeSigma = ThreeSigmaThreshold/TotalCounts*ThreeSigmaRecordedPhotons
    NPhotonsFiveSigma = FiveSigmaThreshold/TotalCounts*FiveSigmaRecordedPhotons
# Differential sensitivity in units of (units of ergs/(cm^2*s))
    ThreeSigmaDiffSensitivity = NPhotonsThreeSigma*SourceEnergy*MeVtoErgs/(Denom)
    FiveSigmaDiffSensitivity = NPhotonsFiveSigma*SourceEnergy*MeVtoErgs/(Denom)
# Differential Sensitivity in units of photons/cm^2/s
    ThreeSigmaPhotonSensitivity = NPhotonsThreeSigma/(Denom)
    FiveSigmaPhotonSensitivity = NPhotonsFiveSigma/(Denom)
    print("At a source energy of " + str(SourceEnergy) + " MeV and exposure time of " + str(EffectiveTime) + " seconds , the sensitivity is (in ergs*cm^-2*s^-1):")
    print("\tThree Sigma: "+str(ThreeSigmaDiffSensitivity))
    print("\tFive Sigma: "+str(FiveSigmaDiffSensitivity))
    print("In photons/cm^2/s:")
    print("\tThree Sigma: "+ str(ThreeSigmaPhotonSensitivity))
    print("\tFive Sigma: "+ str(FiveSigmaPhotonSensitivity))
